handlers/handlers.go

```diff
@@ -141,6 +160,10 @@ func handleAcquire(r *ranch.Ranch) http.HandlerFunc {
 		}
 		logrus.Infof("Resource leased: %v", string(resJSON))
 		fmt.Fprint(res, string(resJSON))
+
+		latency := time.Since(t)
+		labels := prometheus.Labels{"type": rtype, "state": state, "dest": dest, "owner": owner}
+		acquireWaitingDurationSeconds.With(labels).Observe(latency.Seconds())
```

Comment:
This endpoint doesn't block when resources are unavailable, it returns with an error and expects clients to rety, I don't think this change does what you think it does

Reply:
Oh. Oh.

Do you mean this?

https://github.com/kubernetes-sigs/boskos/blob/c0841a578f599dd7330d15a5c77419712fc0ac3e/client/client.go#L413-L416

Client code is a lib.
The app (`ci-op`) which uses the client is not long live process.
Do we need to some kind of push gateway for saving the data in prometheus?

I will ask you tomorrow.

Reply:
We keep the queue of clients in-memory (which is not ideal and makes it impossible to HA `boskos`) so you'd want to use that data store and perhaps improve it to record the time of entry

Reply:
Let me try to do this:

- Clients save the queue in data store (instead of memory): A record in the queue contains when the client started to require the resources.
- The server reads out the record while serving the requirement and it can calculate how much time it is in the queue.

Reply:
@hongkailiu what data store were you thinking about? We should probably model this with a new CRD

Reply:
Yes. Thinking about a CRD and save its instances in etcd.
To replace this:

https://github.com/kubernetes-sigs/boskos/blob/c0841a578f599dd7330d15a5c77419712fc0ac3e/storage/storage.go#L27-L40

Reply:
Maybe the CRD implementation of the storage can be done with another PR?

Reply:
ah. I was somewhat confused by the `Clients save the queue in data store (instead of memory):` because the data store is an in-memory store.

And uhm, IMHO we should keep the changes minimal and not change how the whole queue is implemented, just add the additional info (i.E. return here https://github.com/kubernetes-sigs/boskos/blob/c0841a578f599dd7330d15a5c77419712fc0ac3e/ranch/ranch.go#L117 when the entry was created and pass that info through)